src/WebJobs.Extensions.DurableTask/Listener/FunctionExecutionHelper.cs

```diff
@@ -0,0 +1,144 @@
+                            result.Exception);
+                    }
+                }
+                else
+                {
+                    return new WrappedFunctionResult(
+                        WrappedFunctionResult.FunctionResultStatus.Success,
+                        null);
+                }
+            }
+            catch (Exception e)
+            {
+                return new WrappedFunctionResult(
+                     WrappedFunctionResult.FunctionResultStatus.FunctionsRuntimeError,
+                     e);
+            }
+        }
+
+        public static async Task<WrappedFunctionResult> ExecuteFunction(
+            ITriggeredFunctionExecutor executor,
+            IApplicationLifetimeWrapper hostServiceLifetime,
+            TriggeredFunctionData triggerInput,
+            CancellationToken cancellationToken)
+        {
+#pragma warning disable CS0618 // InvokeHandler approved for use by this extension
+            //if (triggerInput.InvokeHandler != null)
+            //{
+            //    // Activity functions cannot use InvokeHandler, because the usage of InvokeHandler prevents the function from
+            //    // returning a value in the way that the Activity shim knows how to handle.
+            //    throw new ArgumentException(
+            //        $"{nameof(ExecuteFunction)} cannot be used when ${nameof(triggerInput)} has a value for ${nameof(TriggeredFunctionData.InvokeHandler)}");
+            //}
+#pragma warning restore CS0618
+
+            try
+            {
+                FunctionResult result = await executor.TryExecuteAsync(triggerInput, cancellationToken);
+
+                if (!result.Succeeded)
+                {
+                    // This is a best effort approach to determine if this is caused by an exception in
+                    // the webjobs pipeline. False positives are alright, as we still only
+                    // will abort the session when the host is performing a shutdown,
+                    // so it will just fail normally on the second execution.
+                    if (IsHostStopping(hostServiceLifetime) &&
+                        IsHostRelatedException(result.Exception))
+                    {
+                        return new WrappedFunctionResult(
+                             WrappedFunctionResult.FunctionResultStatus.FunctionsRuntimeError,
+                             result.Exception);
+                    }
+                    else
+                    {
+                        return new WrappedFunctionResult(
+                            WrappedFunctionResult.FunctionResultStatus.UserCodeError,
+                            result.Exception);
+                    }
+                }
+                else
+                {
+                    return new WrappedFunctionResult(WrappedFunctionResult.FunctionResultStatus.Success, null);
+                }
+            }
```

Comment:
See my previous comment for a suggestions about how we could make this code easier to read.